File: Session13/C++/TopKfrequentElements.cpp
https://leetcode.com/problems/top-k-frequent-elements/

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        
        priority_queue<int, vector<int>, greater<int>> pq;
        unordered_map<int, int> freq;
        
        for(auto i:nums)
        {
            freq[i]++;
        }
        
        for(auto i:freq)
        {
            pq.push(i.second);
            
            if(pq.size() > k)
            {
                pq.pop();
            }
        }
        
        vector<int> res;
        
        for(auto i:freq)
        {
            if(i.second >= pq.top())
            {
                res.push_back(i.first);
            }
        }
        
        return res;
        
    }
};

File: Session13/Java/TopKfrequentElements.java
https://leetcode.com/problems/top-k-frequent-elements/

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        
        Map<Integer, Integer> map = new HashMap<>();
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(int i: nums)
        {
            map.put(i, map.getOrDefault(i,0)+1);
        }
        
        for (Integer value : map.values()) 
        {
            pq.offer(value);
            if(pq.size()>k)
            {
                pq.poll();
            }
        }
        
        int[] res = new int[k];
        int i=0;
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) 
        {
            Integer key = entry.getKey();
            Integer value = entry.getValue();
            if(value >= pq.peek())
            {
                res[i]=key;
                i++;
            }
        }
        
        return res;
        
    }
}